api/error_unexpected_response_test.go

```diff
@@ -0,0 +1,268 @@
+		ts.Close()
+	})
+
+	// Test the server
+	tc := ts.Client()
+	resp, err := tc.Get(func() string { p, _ := url.JoinPath(ts.URL, "/"); return p }())
+	must.NoError(t, err)
+	defer resp.Body.Close()
+	b, err := io.ReadAll(resp.Body)
+	must.NoError(t, err)
+	t.Logf("checking testServer, got resp: %s", b)
+
+	// If we get here, the testServer is running and ready for requests.
+	return ts.URL
+}
+
+// addMockHeaders sets the common Nomad headers to values sufficient to be
+// parsed into api.QueryMeta
+func addMockHeaders(h http.Header) {
+	h.Add("X-Nomad-Knownleader", "true")
+	h.Add("X-Nomad-Lastcontact", "0")
+	h.Add("X-Nomad-Index", "1")
+	h.Add("Content-Type", "application/json")
+}
+
+// testNotFoundHandler creates a testHandler preconfigured with status code 404.
+func testNotFoundHandler(b string) http.Handler { return testHandler(http.StatusNotFound, b) }
+
+// testNotFoundHandler creates a testHandler preconfigured with status code 200.
+func testOKHandler(b string) http.Handler { return testHandler(http.StatusOK, b) }
+
+// testHandler is a helper function that writes a Nomad-like server response
+// with the necessary headers to make the API client happy
+func testHandler(sc int, b string) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		addMockHeaders(w.Header())
+		w.WriteHeader(sc)
+		w.Write([]byte(b))
+	})
+}
+
+// closingHandler is a handler that terminates the response body early in the
+// reading process
+func closingHandler(sc int, b string) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+		// We need a misbehaving reader to test network effects when collecting
+		// the http.Response data into a UnexpectedResponseError
+		er := iotest.TimeoutReader( // TimeoutReader throws an error on the second read
+			iotest.OneByteReader( // OneByteReader yields a byte at a time, causing multiple reads
+				strings.NewReader(mockNamespaceBody),
+			),
+		)
+
+		// We need to set content-length to the true value it _should_ be so the
+		// API-side reader knows it's a short read.
+		w.Header().Set("content-length", fmt.Sprint(len(mockNamespaceBody)))
+		addMockHeaders(w.Header())
+		w.WriteHeader(sc)
+
+		// Using io.Copy to send the data into w prevents golang from setting the
+		// content-length itself.
+		io.Copy(w, er)
+	})
+}
+
+// testLogRequestHandler wraps a http.Handler with a logger that writes to the
+// test log output
+func testLogRequestHandler(t *testing.T, h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		t := t
+		// call the original http.Handler wrapped in a httpsnoop
+		m := httpsnoop.CaptureMetrics(h, w, r)
+		ri := HTTPReqInfo{
+			uri:       r.URL.String(),
+			method:    r.Method,
+			ipaddr:    ipAddrFromRemoteAddr(r.RemoteAddr),
+			code:      m.Code,
+			duration:  m.Duration,
+			size:      m.Written,
+			userAgent: r.UserAgent(),
+		}
+		t.Logf(ri.String())
+	})
+}
```

Comment:
It doesn't seem like we assert on any of the metrics we're getting from `httpsnoop`, but there's enough indirection that I'm having trouble knowing for sure. Are we adding this wrapper just to add log lines to the test?

Reply:
Yep, it is only for that mock test server to make a basic request log. It's definitely not super critical, but it did help me validate that my test mux was routing as it should have been.